File: wwwroot/js/utils/muscleMapper.js
/**
 * Muscle Mapping Utility
 * Maps muscle names to API muscle group names (synced with C# MuscleMapper)
 */

// Special muscle names that need mapping
const MUSCLE_MAPPING = {
    'hamstrings': 'hamstring',
    'glutes': 'gluteus',
    'upperback': 'back_upper', 
    'lowerback': 'back_lower',
    'backupper': 'back_upper',
    'backlower': 'back_lower',
    'back': 'latissimus',
    'lats': 'latissimus',
    'calves': 'calfs',
    'abdominals': 'abs',
    'traps': 'back_upper',
    'trapezius': 'back_upper',
    'corelower': 'core_lower',
    'coreupper': 'core_upper',
    'shouldersback': 'shoulders_back',
    'shouldersfront': 'shoulders_front',
    'shoulderback': 'shoulders_back',
    'shoulderfront': 'shoulders_front',
    // Extra aliases for AI responses
    'quads': 'quadriceps',
    'pecs': 'chest',
    'pectorals': 'chest',
    'delts': 'shoulders',
    'deltoids': 'shoulders'
};

// Valid API muscle groups
const VALID_MUSCLE_GROUPS = new Set([
    'quadriceps', 'hamstring', 'gluteus', 'chest', 'shoulders', 'triceps', 'biceps',
    'back_upper', 'back_lower', 'back', 'latissimus', 'calfs', 'abs', 'core', 'core_upper', 'core_lower',
    'shoulders_front', 'shoulders_back', 'forearms', 'neck', 'hands', 'legs', 'adductors', 'abductors',
    'all', 'all_upper', 'all_lower'
]);

/**
 * Map muscle names to valid API muscle groups
 * {string[]} muscles - Array of muscle names
 * returns {string[]} Array of valid API muscle group names
 */
function mapMuscleNamesToGroups(muscles) {
    const result = [];
    
    muscles.forEach(muscle => {
        // Normalize: lowercase, trim, remove spaces and underscores
        const normalized = muscle.toLowerCase().trim().replace(/[\s_]/g, '');
        
        // Check if needs mapping (hamstrings -> hamstring, etc)
        if (MUSCLE_MAPPING[normalized]) {
            result.push(MUSCLE_MAPPING[normalized]);
        }
        // Try as-is if it's valid
        else if (VALID_MUSCLE_GROUPS.has(normalized)) {
            result.push(normalized);
        }
        // Otherwise ignore (filters out "OTHER" and invalid muscles)
    });
    
    // Remove duplicates
    return [...new Set(result)];
}

/**
 * Generate muscle visualization image URL
 * {string[]} muscles - Array of muscle names
 * {string} color - RGB color string (default: primary purple)
 * returns {string} Image URL for muscle visualization
 */
function generateMuscleVisualizationUrl(muscles, color = '131,115,218') {
    const validMuscleGroups = mapMuscleNamesToGroups(muscles);
    const muscleGroups = validMuscleGroups.join(',');
    return `/api/musclegroup/image?muscleGroups=${encodeURIComponent(muscleGroups)}&color=${encodeURIComponent(color)}&transparentBackground=true`;
}

File: Business/Helpers/MuscleMapper.cs
namespace WorkoutProgramBuilder.Business.Helpers;


// Backend mapping för server-side rendering
// Helper class for mapping muscle names to API muscle group names
public static class MuscleMapper
{
    // Map special muscle names to API names (normalized: no spaces/underscores)
    private static readonly Dictionary<string, string> MuscleMapping = new()
    {
        { "HAMSTRINGS", "hamstring" },
        { "GLUTES", "gluteus" },
        { "BACK", "latissimus" },
        { "UPPERBACK", "back_upper" },  
        { "LOWERBACK", "back_lower" },
        { "BACKUPPER", "back_upper" },
        { "BACKLOWER", "back_lower" },
        { "LATS", "latissimus" },
        { "CALVES", "calfs" },
        { "ABDOMINALS", "abs" },
        { "TRAPS", "back_upper" },
        { "TRAPEZIUS", "back_upper" },
        { "CORE", "core" },
        { "CORELOWER", "core_lower" },
        { "COREUPPER", "core_upper" },
        { "SHOULDERSBACK", "shoulders_back" },
        { "SHOULDERSFRONT", "shoulders_front" },
        { "SHOULDERBACK", "shoulders_back" },
        { "SHOULDERFRONT", "shoulders_front" }
    };

    // Special muscle groups that expand to multiple muscles
    private static readonly Dictionary<string, List<string>> SpecialGroups = new()
    {
        { "ALL", ["quadriceps", "hamstring", "gluteus", "chest", "shoulders", "triceps", "biceps", "back_upper", "back_lower", "latissimus", "calfs", "abs", "forearms"] },
        { "ALL_UPPER", ["chest", "shoulders", "triceps", "biceps", "back_upper", "latissimus", "abs", "forearms"] },
        { "ALL_LOWER", ["quadriceps", "hamstring", "gluteus", "calfs", "adductors", "abductors"] }
    };
    
    
    // Valid API muscle groups
    private static readonly HashSet<string> ValidMuscleGroups = new()
    {
        "quadriceps", "hamstring", "gluteus", "chest", "shoulders", "triceps", "biceps",
        "back_upper", "back_lower", "back", "latissimus", "calfs", "abs", "core", "core_upper", "core_lower",
        "shoulders_front", "shoulders_back", "forearms", "neck", "hands", "legs", "adductors", "abductors",
        "all", "all_upper", "all_lower"
    };

    // Map muscle names to valid API muscle group names
    public static List<string> MapToApiMuscleGroups(List<string> muscles)
    {
        var result = new List<string>();

        foreach (var muscle in muscles)
        {
            // Normalize: uppercase, trim, remove spaces and underscores
            var normalized = muscle.ToUpper().Trim().Replace(" ", "").Replace("_", "");
            
            // Check if it's a special group (ALL, ALL_UPPER, ALL_LOWER)
            if (SpecialGroups.TryGetValue(normalized, out var specialMuscles))
            {
                result.AddRange(specialMuscles);
            }
            // Check if needs mapping (HAMSTRINGS -> hamstring, etc)
            else if (MuscleMapping.TryGetValue(normalized, out var mapped))
            {
                result.Add(mapped);
            }
            else
            {
                // Try lowercase as-is (chest, biceps, etc)
                var lower = muscle.ToLower().Trim().Replace(" ", "").Replace("_", "");
                if (ValidMuscleGroups.Contains(lower))
                {
                    result.Add(lower);
                }
            }
        }

        return result.Where(m => !string.IsNullOrEmpty(m)).Distinct().ToList();
    }


    // Generate muscle visualization image URL
    public static string GenerateMuscleVisualizationUrl(List<string> muscles, string color = "131,115,218")
    {
        var mappedMuscles = MapToApiMuscleGroups(muscles);
        var muscleGroupsParam = string.Join(",", mappedMuscles);
        return $"/api/musclegroup/image?muscleGroups={Uri.EscapeDataString(muscleGroupsParam)}&color={Uri.EscapeDataString(color)}&transparentBackground=true";
    }
}
